paasta_tools/kubernetes_tools.py

```diff
@@ -486,12 +495,40 @@ def get_autoscaling_params(self) -> AutoscalingParamsDict:
             defaults=default_params,
         )
 
+    # TODO: move the default scaling policy to system paasta configs
+    # TODO: allow per-instance overrides in soa-configs, and merge this function
+    #   with `get_autoscaling_params`
+    def get_autoscaling_scaling_policy(self, max_replicas: int) -> Dict:
+        """Returns the k8s HPA scaling policy in raw JSON. Requires k8s v1.18
+        to work.
+        """
+        # policy in human terms: scale down at most every 60 seconds by 30% of
+        # current replicas or 5% of max replicas, whichever is higher.
+        return {
+            "scaleDown": {
+                "selectPolicy": "Max",
+                "stabilizationWindowSeconds": 300,
```

Comment:
How does the stabilization window interact with the period seconds on the policies? Would it be easier to reason about if they were the same?

Reply:
I've added more detailed comments to explain the difference. They serve two different purposes, and I don't think it makes sense to make them identical. The purposes of the two numbers are:

- `periodSeconds`: Let's say the HPA decides to scale down by a max of 4 pods. Setting `periodSeconds` of 60 seconds of tells the HPA to scale down by a max of 4 pods every 60 seconds. Not setting `periodSeconds` means the HPA will effectively default the `periodSeconds` to the interval at which the HPA runs, which is the HPA sync period. By default, this is 15 seconds.
- `stabilizationWindowSeconds`: Every sync period, the HPA will produce and record a recommended desired replica count. However, if the underlying metrics are fluctuating, the count can too. To fight this, a stabilization window can be set, which is the amount of time in the past the HPA will look for previous desired replica counts, and select the highest one. In this case, the HPA will select the highest desired replica count from the last 300 seconds. The idea is to smoothen downscaling against sometimes unreliable metrics.

Reply:
This question was also coming from the perspective of making it as easy as possible to explain what the HPA is doing in a particular situation.

After trying to reason through both cases, I agree it doesn't really make it easier to explain if they're the same value, so that's fine.